src/datasets/arrow_dataset.py

```diff
@@ -198,27 +198,34 @@ class TensorflowDatasetMixin:
     _TF_DATASET_REFS = set()
 
     @staticmethod
-    def _get_output_signature(dataset, cols_to_retain, test_batch, batch_size):
+    def _get_output_signature(dataset, collate_fn, collate_fn_args, batch_size):
         if config.TF_AVAILABLE:
             import tensorflow as tf
         else:
             raise ImportError("Called a Tensorflow-specific function but Tensorflow is not installed.")
 
+        # Tensorflow needs an exact signature for tf.numpy_function, so
+        # we need to figure out what's coming back in advance. The only way to do this is to run a test batch -
+        # the collator may add columns, so we can't figure it out just by inspecting the dataset.
+        test_batch = dataset[0, 1, 2, 3]
+        test_batch = [{key: value[i] for key, value in test_batch.items()} for i in range(4)]
+        test_batch = collate_fn(test_batch, **collate_fn_args)
+        columns_to_dtypes = dict()
```

Comment:
Nit (for consistency):
```suggestion
        columns_to_dtypes = {}
```